### PR TITLE
Add support for Openstack Designate as a DNS provider 

### PR DESCRIPTION
**Pull request by hsmade**
This PR adds support for Openstack designate as a DNS provider.

### CODE REVIEW

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:
- [ ] add a description to your PR
- [x] be able to maintain this provider
- [x] have a homogeneous design with the other providers
- [x] add tests ("live" and units)
```shell
make test
```
- [x] add [documentation](https://github.com/xenolf/lego/blob/master/CONTRIBUTING.md#dns-providers-api-references)
- [ ] add CLI DNS help: https://github.com/xenolf/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/cmd/cmd_dnshelp.go#L19
- [ ] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf ./lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -s https://acme-staging-v02.api.letsencrypt.org/directory -d *.example.com -d example.com run
```
- [x] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```

**Comment by hsmade** [comment]
```wifournier@worker001:~$ . .dev-wifournier-openrc.sh 
Please enter your OpenStack Password: 
wifournier@worker001:~$ ./lego --server=https://acme-staging-v02.api.letsencrypt.org/directory --email=wifournier@xxxx.xxx --domains="test-lego.dev-wifournier.xxxx" --dns designate  --dns.disable-cp run   
2019/02/06 13:57:11 [INFO] [test-lego.dev-wifournier.xxxx] acme: Obtaining bundled SAN certificate
2019/02/06 13:57:12 [INFO] [test-lego.dev-wifournier.xxxx] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/xxxxxxxxxxxxxxxxxxx
2019/02/06 13:57:12 [INFO] [test-lego.dev-wifournier.xxxx] acme: Could not find solver for: tls-alpn-01
2019/02/06 13:57:12 [INFO] [test-lego.dev-wifournier.xxxx] acme: Could not find solver for: http-01
2019/02/06 13:57:12 [INFO] [test-lego.dev-wifournier.xxxx] acme: use dns-01 solver
2019/02/06 13:57:12 [INFO] [test-lego.dev-wifournier.xxxx] acme: Preparing to solve DNS-01
2019/02/06 13:57:13 [INFO] [test-lego.dev-wifournier.xxxx] acme: Trying to solve DNS-01
2019/02/06 13:57:13 [INFO] [test-lego.dev-wifournier.xxxx] acme: Checking DNS record propagation using [xxxx.xxx.:53]
2019/02/06 13:57:13 [INFO] Wait for propagation [timeout: 10m0s, interval: 10s]
2019/02/06 13:57:19 [INFO] [test-lego.dev-wifournier.xxxx] The server validated our request
2019/02/06 13:57:19 [INFO] [test-lego.dev-wifournier.xxxx] acme: Cleaning DNS-01 challenge
2019/02/06 13:57:19 [INFO] [test-lego.dev-wifournier.xxxx] acme: Validations succeeded; requesting certificates
2019/02/06 13:57:21 [INFO] [test-lego.dev-wifournier.xxxx] Server responded with a certificate.
```

**Comment by ldez** [comment]
The purpose of the command is to validate the generation of wildcard certificates.

So could you re-do the command:
```console
$ rm -rf .lego

$ ./lego -m your@email.com --dns designate -s https://acme-staging-v02.api.letsencrypt.org/directory -d *.dev-wifournier.xxxx -d dev-wifournier.xxxx run
```

**Comment by ldez** [comment]
Also the CLI DNS help is not complete.

**Comment by hsmade** [comment]
```wifournier@worker001:~$ ./lego --server=https://acme-staging-v02.api.letsencrypt.org/directory --email=wifournier@ebay.com --domains="*.test-lego.dev-wifournier.xxxx" --domains="test-lego.dev-wifournier.xxxx" --dns designate  --dns.disable-cp run   
2019/02/06 14:11:17 [INFO] [*.test-lego.dev-wifournier.xxxx, test-lego.dev-wifournier.xxxx] acme: Obtaining bundled SAN certificate
2019/02/06 14:11:18 [INFO] [*.test-lego.dev-wifournier.xxxx] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/xxxxxxxxxxxxxxxxx
2019/02/06 14:11:18 [INFO] [test-lego.dev-wifournier.xxxx] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/xxxxxxxxxxxxxxxxx
2019/02/06 14:11:18 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: authorization already valid; skipping challenge
2019/02/06 14:11:18 [INFO] [test-lego.dev-wifournier.xxxx] acme: Could not find solver for: tls-alpn-01
2019/02/06 14:11:18 [INFO] [test-lego.dev-wifournier.xxxx] acme: Could not find solver for: http-01
2019/02/06 14:11:18 [INFO] [test-lego.dev-wifournier.xxxx] acme: use dns-01 solver
2019/02/06 14:11:18 [INFO] [test-lego.dev-wifournier.xxxx] acme: Preparing to solve DNS-01
2019/02/06 14:11:19 [INFO] [test-lego.dev-wifournier.xxxx] acme: Trying to solve DNS-01
2019/02/06 14:11:19 [INFO] [test-lego.dev-wifournier.xxxx] acme: Checking DNS record propagation using [127.0.0.1:53]
2019/02/06 14:11:19 [INFO] Wait for propagation [timeout: 10m0s, interval: 10s]
2019/02/06 14:11:24 [INFO] [test-lego.dev-wifournier.xxxx] The server validated our request
2019/02/06 14:11:24 [INFO] [test-lego.dev-wifournier.xxxx] acme: Cleaning DNS-01 challenge
2019/02/06 14:11:25 [INFO] [*.test-lego.dev-wifournier.xxxx, test-lego.dev-wifournier.xxxx] acme: Validations succeeded; requesting certificates
2019/02/06 14:11:28 [INFO] [*.test-lego.dev-wifournier.xxxx] Server responded with a certificate.

```

**Comment by ldez** [comment]
> 2019/02/06 14:11:18 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: authorization already valid; skipping challenge

Could you do a `rm -rf .lego` before the run of the command?

**Comment by hsmade** [comment]
``` rm -rf .lego/; ./lego -s https://acme-staging-v02.api.letsencrypt.org/directory -m wifournier@xxxx -d "*.test-lego.dev-wifournier.xxxx" -d "test-lego.dev-wifournier.xxxx" --dns designate  --dns.disable-cp -a run   
2019/02/06 15:08:04 No key found for account wifournier@xxxx. Generating a curve P384 EC key.
2019/02/06 15:08:04 Saved key to /media/home/wifournier/.lego/accounts/acme-staging-v02.api.letsencrypt.org/wifournier@xxxx/keys/wifournier@xxxx.key
2019/02/06 15:08:04 [INFO] acme: Registering account for wifournier@xxxx
!!!! HEADS UP !!!!

                Your account credentials have been saved in your Let's Encrypt
                configuration directory at "/media/home/wifournier/.lego/accounts".
                You should make a secure backup of this folder now. This
                configuration directory will also contain certificates and
                private keys obtained from Let's Encrypt so making regular
                backups of this folder is ideal.2019/02/06 15:08:05 [INFO] [*.test-lego.dev-wifournier.xxxx, test-lego.dev-wifournier.xxxx] acme: Obtaining bundled SAN certificate
2019/02/06 15:08:06 [INFO] [*.test-lego.dev-wifournier.xxxx] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/xxxxxxxxxxxxxxxxxxx
2019/02/06 15:08:06 [INFO] [test-lego.dev-wifournier.xxxx] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/xxxxxxxxxxxxxxxxxxx
2019/02/06 15:08:06 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: use dns-01 solver
2019/02/06 15:08:06 [INFO] [test-lego.dev-wifournier.xxxx] acme: Could not find solver for: tls-alpn-01
2019/02/06 15:08:06 [INFO] [test-lego.dev-wifournier.xxxx] acme: Could not find solver for: http-01
2019/02/06 15:08:06 [INFO] [test-lego.dev-wifournier.xxxx] acme: use dns-01 solver
2019/02/06 15:08:06 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: Preparing to solve DNS-01
2019/02/06 15:08:06 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: Trying to solve DNS-01
2019/02/06 15:08:06 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: Checking DNS record propagation using [127.0.0.1:53]
2019/02/06 15:08:06 [INFO] Wait for propagation [timeout: 10m0s, interval: 10s]
2019/02/06 15:08:12 [INFO] [*.test-lego.dev-wifournier.xxxx] The server validated our request
2019/02/06 15:08:12 [INFO] [*.test-lego.dev-wifournier.xxxx] acme: Cleaning DNS-01 challenge
2019/02/06 15:08:13 [INFO] sequence: wait for 1m0s
2019/02/06 15:09:13 [INFO] [test-lego.dev-wifournier.xxxx] acme: Preparing to solve DNS-01
2019/02/06 15:09:14 [INFO] [test-lego.dev-wifournier.xxxx] acme: Trying to solve DNS-01
2019/02/06 15:09:14 [INFO] [test-lego.dev-wifournier.xxxx] acme: Checking DNS record propagation using [127.0.0.1:53]
2019/02/06 15:09:14 [INFO] Wait for propagation [timeout: 10m0s, interval: 10s]
2019/02/06 15:09:19 [INFO] [test-lego.dev-wifournier.xxxx] The server validated our request
2019/02/06 15:09:19 [INFO] [test-lego.dev-wifournier.xxxx] acme: Cleaning DNS-01 challenge
2019/02/06 15:09:20 [INFO] [*.test-lego.dev-wifournier.xxxx, test-lego.dev-wifournier.xxxx] acme: Validations succeeded; requesting certificates
2019/02/06 15:09:21 [INFO] [*.test-lego.dev-wifournier.xxxx] Server responded with a certificate.
```

File: providers/dns/designate/designate.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		return
+	}
+
+	for _, zone := range allZones {
+		if zone.Name == wanted {
+			zoneID = zone.ID
+			return
+		}
+	}
+	return
```
Comment (ldez) [inline providers/dns/designate/designate.go]: if the zone ID is not found then an error must be returned.

File: providers/dns/designate/designate.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	if err != nil {
+		return err
+	}
+
+	zoneID, err := getZoneID(d.client, authZone)
+	if err != nil {
+		return fmt.Errorf("designate: couldn't get zone ID in CleanUp: %sv", err)
+	}
+
+	// use mutex to prevent race condition from GetInfo until SetDNSEntries
```
Comment (ldez) [inline providers/dns/designate/designate.go]: wrong copy/paste :smile:

File: providers/dns/designate/designate.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+		return
+	}
+
+	for _, record := range allRecords {
+		if record.Name == wanted {
+			recordID = record.ID
+			return
+		}
+	}
+	return
```
Comment (ldez) [inline providers/dns/designate/designate.go]: if the record ID is not found then an error must be returned.
Or you need to validate to recordID when you are using `getRecordID`

File: providers/dns/designate/designate.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+	createOpts := recordsets.CreateOpts{
+		Name:        fqdn,
+		Type:        "TXT",
+		TTL:         d.config.TTL,
+		Description: "ACME verification record",
+		Records:     []string{value},
+	}
+
+	// use mutex to prevent race condition from GetInfo until SetDNSEntries
```
Comment (ldez) [inline providers/dns/designate/designate.go]: wrong copy/paste :smile:

**Pull request merged**